#! /bin/sh /usr/share/dpatch/dpatch-run
## battery_acpi_complain.dpatch by Sebastian Harl <sh@tokkee.org>
##
## DP: battery plugin: Don't complain about missing /proc/acpi/battery each
## DP: interval. Instead, complain once when the read callback is called for
## DP: the first time.

@DPATCH@

diff a/src/battery.c b/src/battery.c
--- a/src/battery.c
+++ b/src/battery.c
@@ -23,6 +23,8 @@
 #include "common.h"
 #include "plugin.h"
 
+#include "utils_complain.h"
+
 #if HAVE_MACH_MACH_TYPES_H
 #  include <mach/mach_types.h>
 #endif
@@ -436,6 +438,8 @@ static int battery_read (void)
 /* #endif HAVE_IOKIT_IOKITLIB_H || HAVE_IOKIT_PS_IOPOWERSOURCES_H */
 
 #elif KERNEL_LINUX
+	static c_complain_t acpi_dir_complaint = C_COMPLAIN_INIT_STATIC;
+
 	FILE *fh;
 	char buffer[1024];
 	char filename[256];
@@ -506,8 +510,17 @@ static int battery_read (void)
 			battery_submit ("0", "voltage", voltage);
 	}
 
-	walk_directory (battery_acpi_dir, battery_read_acpi,
-			/* user_data = */ NULL);
+	if (0 == access (battery_acpi_dir, R_OK))
+		walk_directory (battery_acpi_dir, battery_read_acpi,
+				/* user_data = */ NULL);
+	else
+	{
+		char errbuf[1024];
+		c_complain_once (LOG_WARNING, &acpi_dir_complaint,
+				"battery plugin: Failed to access `%s': %s",
+				battery_acpi_dir,
+				sstrerror (errno, errbuf, sizeof (errbuf)));
+	}
 
 #endif /* KERNEL_LINUX */
 
